File: srm626/FixedDiceGameDiv2.py
#! -*- coding:utf-8 -*-
"""
1e-9のような値を答えるときはPythonだと計算が合わなくなる可能性あり
440 / 5.0 =  7.999999999999999 (not 8.0)
ただし、32bitと64bitでも結果が異なる
"""


class FixedDiceGameDiv2:
    def getExpectation(self, a, b):
        # 小数で宣言する
        count = 0.0
        rv = 0.0
        for i in range(1, b + 1):
            for j in range(i + 1, a + 1):
                rv += j
                count += 1
        print ("%s / %s" % (rv, count))
        return rv / count
data = [
    [2, 2],
    [4, 2],
    [3, 3],
    [11, 13]
]

for d in data:
    f = FixedDiceGameDiv2()
    print(f.getExpectation(*d))

File: srm626/FixedDiceGameDiv2.cpp
#include <iostream>
#include <iomanip>
#include <string>
using namespace std;

class FixedDiceGameDiv2 {
public:
  double getExpectation(int a, int b){
    int count = 0;
    int rv = 0;
    for (int i = 1; i < b + 1; i++){
      for (int j = i + 1; j < a + 1; j++){
        rv += j;
        count++;
      }
    }
    return rv / (double)count;
  }
};
  
int main(int argc, char* argv[]){
  FixedDiceGameDiv2 f;
  int a[4][2] = {
    {2, 2},
    {4, 2},
    {3, 3},
    {11, 13},
  };
  for(int i = 0 ; i < 4; i++){
    printf("%lf\n", f.getExpectation(a[i][0], a[i][1]));
  }
}
